akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala

```diff
@@ -0,0 +1,1577 @@
+ *   of changes, if any.
+ * @param maxDeltaElements Maximum number of entries to transfer in one
+ *   gossip message when synchronizing the replicas. Next chunk will be
+ *   transferred in next round of gossip.
+ * @param pruningInterval How often the Replicator checks for pruning of
+ *   data associated with removed cluster nodes.
+ * @param maxPruningDissemination How long time it takes (worst case) to spread
+ *   the data to all other replica nodes. This is used when initiating and
+ *   completing the pruning process of data associated with removed cluster nodes.
+ *   The time measurement is stopped when any replica is unreachable, so it should
+ *   be configured to worst case in a healthy cluster.
+ */
+final class ReplicatorSettings(
+  val role: Option[String],
+  val gossipInterval: FiniteDuration,
+  val notifySubscribersInterval: FiniteDuration,
+  val maxDeltaElements: Int,
+  val pruningInterval: FiniteDuration,
+  val maxPruningDissemination: FiniteDuration) {
+
+  def withRole(role: String): ReplicatorSettings = copy(role = ReplicatorSettings.roleOption(role))
+
+  def withRole(role: Option[String]): ReplicatorSettings = copy(role = role)
+
+  def withGossipInterval(gossipInterval: FiniteDuration): ReplicatorSettings =
+    copy(gossipInterval = gossipInterval)
+
+  def withNotifySubscribersInterval(notifySubscribersInterval: FiniteDuration): ReplicatorSettings =
+    copy(notifySubscribersInterval = notifySubscribersInterval)
+
+  def withMaxDeltaElements(maxDeltaElements: Int): ReplicatorSettings =
+    copy(maxDeltaElements = maxDeltaElements)
+
+  def withPruning(pruningInterval: FiniteDuration, maxPruningDissemination: FiniteDuration): ReplicatorSettings =
+    copy(pruningInterval = pruningInterval, maxPruningDissemination = maxPruningDissemination)
+
+  private def copy(
+    role: Option[String] = role,
+    gossipInterval: FiniteDuration = gossipInterval,
+    notifySubscribersInterval: FiniteDuration = notifySubscribersInterval,
+    maxDeltaElements: Int = maxDeltaElements,
+    pruningInterval: FiniteDuration = pruningInterval,
+    maxPruningDissemination: FiniteDuration = maxPruningDissemination): ReplicatorSettings =
+    new ReplicatorSettings(role, gossipInterval, notifySubscribersInterval, maxDeltaElements, pruningInterval, maxPruningDissemination)
+}
+
+object Replicator {
+
+  /**
+   * Factory method for the [[akka.actor.Props]] of the [[Replicator]] actor.
+   */
+  def props(settings: ReplicatorSettings): Props =
+    Props(new Replicator(settings)).withDeploy(Deploy.local)
+
+  sealed trait ReadConsistency {
+    def timeout: FiniteDuration
+  }
+  case object ReadLocal extends ReadConsistency {
+    override def timeout: FiniteDuration = Duration.Zero
+  }
+  final case class ReadFrom(n: Int, timeout: FiniteDuration) extends ReadConsistency {
+    require(n >= 2, "ReadFrom n must be >= 2, use ReadLocal for n=1")
+  }
+  final case class ReadQuorum(timeout: FiniteDuration) extends ReadConsistency
```

Comment:
Thinking about the discussion we had the other day, this should probably be called `ReadMajority`, since quorum has a more statically defined connotation.